TB3 example added

rjshim opened the pull request:
Added the following updates + modifications.

1. Action files are added to [turtlebot3_msgs repository](https://github.com/ROBOTIS-GIT/turtlebot3_msgs/pull/19)
2. TB3 example added to the list in the package.xml file in the meta package 
3. TB3 examples(i.e., obstacle detection, patrol and position control) rewritten based on the [files](https://github.com/ROBOTIS-GIT/turtlebot3/tree/master/turtlebot3_example/nodes) are added to the repository.

Interactive marker example will be also added when porting work for the interactive maker package is done.

Thank you,
Ryan

robotpilot commented:
@rjshim @routiful 
It passed Travis CI. https://travis-ci.org/ROBOTIS-GIT/turtlebot3/builds/589254734?utm_source=github_status
Please resume PR review.

routiful commented:
@kijongGil 
I think you have to review this code :)

routiful commented on turtlebot3_example/turtlebot3_example/turtlebot3_obstacle_detection/turtlebot3_obstacle_detection.py:
```diff
@@ -0,0 +1,91 @@
+# Authors: Ryan Shim, Gilbert
+
+from geometry_msgs.msg import Twist
+import numpy
```
`numpy` is not included ros2 lib.

rjshim replied:
Can you tell me more about this comment? Are you recommending another library instead of numpy?

routiful replied:
Sorry for my insufficient the comment.
The `numpy` is general library in python3. 
Therefore, you should have import `numpy` before `geometry_msgs` that is general library in ROS.

You can find order of import or include library in robotis code style guide.

rjshim replied:
@routiful 
Thank you.

routiful commented on turtlebot3_example/turtlebot3_example/turtlebot3_obstacle_detection/turtlebot3_obstacle_detection.py:
```diff
@@ -0,0 +1,91 @@
+            qos_profile=qos_profile_sensor_data)
+        self.raw_cmd_vel_sub = self.create_subscription(
+            Twist,
+            'raw_cmd_vel',
```
How about changed this topic name to `cmd_vel_raw`?

rjshim replied:
modified

routiful commented on turtlebot3_example/turtlebot3_example/turtlebot3_obstacle_detection/turtlebot3_obstacle_detection.py:
```diff
@@ -0,0 +1,91 @@
+    def detect_obstacle_callback(self):
+        twist = Twist()
+        obstacle_distance = min(self.scan_ranges)
+        safety_distance = 0.3  # unit: m
```
How about to get this variable to get arguments of main function or ros2 parameter?

rjshim replied:
That can be added when teleop can be run using 'ros2 launch ~' command.

routiful replied:
Then, you should set that variable to use ros parameter service API.
https://github.com/ROBOTIS-Platform/ros2_examples/blob/1e29a5e119dd99e811a29fc66309c10f161be542/examples_rclpy/examples_rclpy/subscriber/observer.py#L56

routiful commented on turtlebot3_example/turtlebot3_example/turtlebot3_position_control/turtlebot3_position_control.py:
```diff
@@ -0,0 +1,180 @@
+    def odom_callback(self, msg):
+        self.last_pose_x = msg.pose.pose.position.x
+        self.last_pose_y = msg.pose.pose.position.y
+        _, _, self.last_pose_theta = self.euler_from_quaternion(msg.pose.pose.orientation)
```
What is `_,` this symbol??

rjshim replied:
I refered to [this](https://mingrammer.com/underscore-in-python/). Let me know if you know any other better expressions.

rjshim commented on turtlebot3_example/turtlebot3_example/turtlebot3_patrol_client/turtlebot3_patrol_client.py:
```diff
@@ -0,0 +1,108 @@
+import sys
+import termios
+
+from action_msgs.msg import GoalStatus
```
modified as you said

rjshim commented on turtlebot3_example/turtlebot3_example/turtlebot3_patrol_server/turtlebot3_patrol_server.py:
```diff
@@ -0,0 +1,127 @@
+#
+# Authors: Ryan Shim, Gilbert
+
+import math
```
modified as you said

rjshim replied:
Would you tell me more about this comment? Are you saying I do not need to import it? (As far as what I googled, I still need to ...?)

kijongGil replied:
Python's basic modules are declare on the top line.
for example
==== basic modules====
import math 
import time

====Third-Party modules ====
from rclpy.action import ActionServer
from rclpy.action import CancelResponse

====local-lrbrary modules ====
from turtlebot3_example.turtlebot3_patrol_server.turtlebot3_path

rjshim replied:
@kijongGil  Thank you for the comment. I got it

rjshim commented on turtlebot3_example/turtlebot3_example/turtlebot3_patrol_server/turtlebot3_patrol_server.py:
```diff
@@ -0,0 +1,127 @@
+from turtlebot3_example.turtlebot3_patrol_server.turtlebot3_path \
+    import Turtlebot3Path
+
+from geometry_msgs.msg import Twist
```
modified as you said

rjshim commented on turtlebot3_example/turtlebot3_example/turtlebot3_patrol_server/turtlebot3_patrol_server.py:
```diff
@@ -0,0 +1,127 @@
+from rclpy.node import Node
+from rclpy.qos import QoSProfile
+import time
+from turtlebot3_example.turtlebot3_patrol_server.turtlebot3_path \
```
modified as you said

kijongGil commented on turtlebot3_example/turtlebot3_example/turtlebot3_obstacle_detection/turtlebot3_obstacle_detection.py:
```diff
@@ -0,0 +1,91 @@
+        ************************************************************"""
+        self.linear_velocity = 0.0  # unit: m/s
+        self.angular_velocity = 0.0  # unit: m/s
+        self.scan_ranges = numpy.ones(360) * numpy.Infinity  # Scan resolution is 1[deg]
```
```suggestion
        self.scan_ranges = []  # Scan resolution is 1[deg]
```
How about this suggestion? Because scan resolution is can be changed by custom tune.

rjshim replied:
Fixed as your suggestion. Thank you

kijongGil commented on turtlebot3_example/turtlebot3_example/turtlebot3_obstacle_detection/main.py:
```diff
@@ -0,0 +1,34 @@
+#
+# Authors: Ryan Shim, Gilbert
+
+import rclpy
```
Please insert enter.

rjshim replied:
Thank you

kijongGil commented on turtlebot3_example/turtlebot3_example/turtlebot3_patrol_client/main.py:
```diff
@@ -0,0 +1,31 @@
+#
+# Authors: Ryan Shim, Gilbert
+
+import rclpy
```
Please insert enter.

rjshim replied:
Thank you

kijongGil commented on turtlebot3_example/turtlebot3_example/turtlebot3_patrol_server/main.py:
```diff
@@ -0,0 +1,34 @@
+#
+# Authors: Ryan Shim, Gilbert
+
+import rclpy
```
Please insert enter.

rjshim replied:
Thank you

kijongGil commented on turtlebot3_example/turtlebot3_example/turtlebot3_position_control/main.py:
```diff
@@ -0,0 +1,34 @@
+#
+# Authors: Ryan Shim, Gilbert
+
+import rclpy
```
Please insert enter.

rjshim replied:
Thank you

kijongGil commented on turtlebot3_example/turtlebot3_example/turtlebot3_position_control/turtlebot3_path.py:
```diff
@@ -0,0 +1,50 @@
+#
+# Authors: Ryan Shim, Gilbert
+
+import math
```
Please insert enter.

rjshim replied:
Thank you

kijongGil commented on turtlebot3_example/turtlebot3_example/turtlebot3_position_control/turtlebot3_position_control.py:
```diff
@@ -0,0 +1,179 @@
+import termios
+from turtlebot3_example.turtlebot3_position_control.turtlebot3_path import Turtlebot3Path
+from geometry_msgs.msg import Twist
+from nav_msgs.msg import Odometry
```
Please regroup modules

rjshim replied:
Thank you for your comment

rjshim commented on turtlebot3_example/setup.py:
```diff
@@ -0,0 +1,48 @@
+    entry_points={
+        'console_scripts': [
+            # To be added
+            # 'turtlebot3_interactive_marker = turtlebot3_example.turtlebot3_interactive_marker.main:main', 
```
This package still needs more debugging work. I will wait until mid Novermber and test it again.

rjshim commented:
I added the following modifications.

1. Reordered 'import modules' 
2. Changed scan_ranges init value 
3. Changed a topic name (raw_cmd_vel -> cmd_vel_raw)

Thank you,
Ryan

routiful commented:
LGTM to me
